lib/web/sessions.go

```diff
@@ -850,18 +851,34 @@ func (s *sessionCache) invalidateSession(ctx context.Context, sctx *SessionConte
 	if err != nil {
 		return trace.Wrap(err)
 	}
+
+	// App session, SAML session and web session deletion should be treated as a single transaction.
+	// To avoid aborting deletion midpoint due to a failure in one of the session deletion,
+	// we use sessionDeletionErr below to join errors and return them at last.
+	var sessionDeletionErr error
```

Comment:
Suggestion:

```suggestion
	var deleteErrs []error
```

and then append to it? This gives us a single, "flat" joined error instead of a joined tree. We can also rely on errors.Join to decide if the result is `nil`.

See the following comments for examples.

Reply:
@codingllama Curios, what would be pros of "flat" error over joined ones? 

> errors.Join to decide if the result is nil

Yeah but we also check for if err is `NotFound` so we aren't doing much.

Reply:
The snippet would go something like this:

```go

	// App session, SAML session and web session deletion should be treated as a single transaction.
	// To avoid aborting deletion midpoint due to a failure in one of the session deletion,
	// we use sessionDeletionErr below to join errors and return them at last.
	var deleteErrs error
	if err := clt.DeleteUserAppSessions(ctx, &proto.DeleteUserAppSessionsRequest{Username: sctx.GetUser()}); err != nil {
		deleteErrs = append(deleteErrs, err)
	}
	if samlSession := sctx.cfg.Session.GetSAMLSession(); samlSession != nil && samlSession.ID != "" {
		if err := clt.DeleteSAMLIdPSession(ctx, types.DeleteSAMLIdPSessionRequest{
			SessionID: samlSession.ID,
		}); err != nil && !trace.IsNotFound(err) {
			deleteErrs = append(deleteErrs, err)
		}
	}
	// Delete just the session - leave the bearer token to linger to avoid
	// failing a client query still using the old token.
	if err := clt.WebSessions().Delete(ctx, types.DeleteWebSessionRequest{
		User:      sctx.GetUser(),
		SessionID: sctx.GetSessionID(),
	}); err != nil && !trace.IsNotFound(err) {
		deleteErrs = append(deleteErrs, err)
	}

	if err := errors.Join(deleteErrs...); err != nil {
		return trace.Wrap(deleteErrs)
	}
```

We could also drop most ifs and always join, like this:

```go

	// App session, SAML session and web session deletion should be treated as a single transaction.
	// To avoid aborting deletion midpoint due to a failure in one of the session deletion,
	// we use sessionDeletionErr below to join errors and return them at last.
	var deleteErrs error
	err = clt.DeleteUserAppSessions(ctx, &proto.DeleteUserAppSessionsRequest{Username: sctx.GetUser()})
	deleteErrs = append(deleteErrs, err)

	if samlSession := sctx.cfg.Session.GetSAMLSession(); samlSession != nil && samlSession.ID != "" {
		if err := clt.DeleteSAMLIdPSession(ctx, types.DeleteSAMLIdPSessionRequest{
			SessionID: samlSession.ID,
		}); err != nil && !trace.IsNotFound(err) {
			deleteErrs = append(deleteErrs, err)
		}
	}

	// Delete just the session - leave the bearer token to linger to avoid
	// failing a client query still using the old token.
	err = clt.WebSessions().Delete(ctx, types.DeleteWebSessionRequest{
		User:      sctx.GetUser(),
		SessionID: sctx.GetSessionID(),
	})
	deleteErrs = append(deleteErrs, err)

	if err := errors.Join(deleteErrs...); err != nil {
		return trace.Wrap(deleteErrs)
	}
```

Reply:
> A "flat" join would traverse with less hops. It's not a big deal.

Thanks for explaining and reference code. In the second example, we can only skip `if` block in the `DeleteUserAppSessions`.  For the first example, Since we are calling three different delete func, as opposed to calling one delete func thrice, I don't think we are gaining much here.

From the `errors.Join()` godoc:
> // The error formats as the concatenation of the strings obtained
> // by calling the Error method of each element of errs, with a newline
> // between each string.